Creeado por: Diego Vera 20172404

Pregunta 1:

Se le brinda dos códigos en C: Programa1_1.c y Programa1_2.c. Ambos generan el mismo
resultado pero se han implementado de diferentes maneras. Se le pide lo siguiente:

1) (1.0 punto) Incluir la medición de tiempo desde C considerando únicamente el
cálculo de la suma para ambos programas. Imprimir el tiempo de ejecución al
ejecutar los programas. Utilice las unidades que mejor se ajusten al tiempo de
ejecución de su PC. No usar más de dos decimales.


Mediciones para Programa1_1:

In [171]:
%%file Programa1_1.c

#include <stdio.h>
#include <time.h> //para tomar mediciones del tiempo

#define N 1000000

int main() {
  int sum = 0;
  int array[N];

  // Initialize the array
  for (int i = 0; i < N; i++) {
    array[i] = i;
  }
    
  struct timespec ti, tf;
  double elapsed;

  clock_gettime(CLOCK_REALTIME, &ti);
  // Calculate the sum of the array
  for (int i = 0; i < N; i++) {
    sum += array[i];
  }
  clock_gettime(CLOCK_REALTIME, &tf);
  
  elapsed = (tf.tv_sec - ti.tv_sec) * 1e9 + (tf.tv_nsec - ti.tv_nsec);
    printf("el tiempo en mili que toma la función en C es %.2lf\n", elapsed*1e-6);
  

  // Print the sum
  printf("Sum: %d\n", sum);

  return 0;
}



Overwriting Programa1_1.c


In [172]:
! gcc Programa1_1.c -o Programa1_1

In [174]:
! ./Programa1_1

el tiempo en mili que toma la función en C es 4.62
Sum: 1783293664


Mediciones en Programa1_2:

In [180]:
%%file Programa1_2.c

#include <stdio.h>
#include <time.h> 

#define N 1000000

int main() {
  int sum = 0;

  struct timespec ti, tf;
  double elapsed;

  clock_gettime(CLOCK_REALTIME, &ti);
  // Calculate the sum of the array
  for (int i = 0; i < N; i++) {
    sum += i;
  }
  clock_gettime(CLOCK_REALTIME, &tf);
  
  elapsed = (tf.tv_sec - ti.tv_sec) * 1e9 + (tf.tv_nsec - ti.tv_nsec);
  printf("el tiempo en mili que toma la función en C es %.2lf\n", elapsed*1e-6);

  // Print the sum
  printf("Sum: %d\n", sum);

  return 0;
}



Overwriting Programa1_2.c


In [181]:
! gcc Programa1_2.c -o Programa1_2

In [182]:
! ./Programa1_2

el tiempo en mili que toma la función en C es 3.69
Sum: 1783293664


2) (1.0 punto) Realizar 15 ejecuciones desde el terminal para ambos archivos
ejecutables. Realizar una captura de pantalla de su ejecución. Los tiempos en su
imagen deben ser visibles.

Para el 1:

In [183]:
#Un for desde jupyter
for i in range (15):
    !./Programa1_1

el tiempo en mili que toma la función en C es 4.31
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.87
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.93
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.91
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.97
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.96
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.86
Sum: 1783293664
el tiempo en mili que toma la función en C es 4.04
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.99
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.95
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.91
Sum: 1783293664
el tiempo en mili que toma la función en C es 4.01
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.92
Sum: 1783293664
el tiempo en mili que toma la función en C es 4.05
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.84
Sum: 178329

Para el 2:

In [184]:
#de manera alternativa con un for desde jupyter
for i in range (15):
    !./Programa1_2

el tiempo en mili que toma la función en C es 3.72
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.75
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.93
Sum: 1783293664
el tiempo en mili que toma la función en C es 4.01
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.92
Sum: 1783293664
el tiempo en mili que toma la función en C es 4.09
Sum: 1783293664
el tiempo en mili que toma la función en C es 4.25
Sum: 1783293664
el tiempo en mili que toma la función en C es 4.22
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.91
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.76
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.94
Sum: 1783293664
el tiempo en mili que toma la función en C es 3.76
Sum: 1783293664
el tiempo en mili que toma la función en C es 4.28
Sum: 1783293664
el tiempo en mili que toma la función en C es 4.11
Sum: 1783293664
el tiempo en mili que toma la función en C es 4.10
Sum: 178329

3) (1.0 punto) Adjuntar una tabla con todas las mediciones.

| Programa1_1 | Programa1_2 |
|:-----------:|:-----------:|
|    4.31     |     3.72    |
|    3.87     |     3.75    |
|    3.93     |     3.93    |
|    3.91     |     4.01    |
|    3.97     |     3.92    |
|    4.96     |     4.09    |
|    3.86     |     4.25    |
|    4.04     |     4.22    |
|    3.99     |     3.91    |
|    3.95     |     3.76    |
|    3.91     |     3.94    |
|    4.01     |     3.76    |
|    3.92     |     4.28    |
|    4.05     |     4.11    |
|    3.84     |     4.10    |

4) (1.0 punto) Explicar cómo se relaciona el Programa1_1.c con los conceptos de
localidad temporal y localidad espacial.

In [185]:
#sacando promedio
(4.31+3.87+3.93+3.91+3.97+4.96+3.86+4.04+3.99+3.95+3.91+4.01+3.92+4.05+3.84)/15

4.034666666666666

Para el programa1_1 calculando el promedio de las 15 tenemos como resultado: 4.035

Analizando el programa vemos que inicialmente en el main se declara la variable suma y el arreglo con n terminos, al declarar el arreglo hacemos mas facil la localidad espacial para los valores del arreglo y mas si los utilizamos los más proximos

Dentro del for se inicializa el arreglo guardandole los valores
Al utilizar el for y guardar valores se hace buena practica de localidad espacial ya que posiciones de memoria cercanos van a ser accedidos ya que se declaro anteriormente 

Entre mas pequeño es el buble , mejor es la localidad espacial


Las mediciones son tomadas unicamente en la funcion que calcula la suma de los valores del termino.

5) (1.0 punto) Explicar cómo se relaciona el Programa1_2.c con los conceptos de
localidad temporal y localidad espacial.


In [186]:
#sacando promedio
(3.72+3.75+3.93+4.01+3.92+4.09+4.25+4.22+3.91+3.76+3.94+3.76+4.28+4.11+4.10)/15

3.983333333333333

El tiempo promedio es de 3.98

Analizando el programa podemos apreciar que directamente se declara la variable suma y se va directamente al bucle para calcular el valor de la suma de N terminos.
Directamente la localidad temporal se es usada ya que al usar una direccion para darle el valor a i , luego nuevamente se accede a esta para guardar su valor en la variable suma principalmente.

6) (1.0 punto) Utilizar el comando getconf -a | grep CACHE en el terminal y verificar el
tamaño de bloque de su computador. Recordar que el tamaño de bloque en el nivel
1 se ve en la línea LEVEL1_DCACHE_LINESIZE (en bytes). Adjuntar captura de
pantalla

In [188]:
#Se utilizo el comando para no adjuntar captura:
!getconf -a | grep CACHE

LEVEL1_ICACHE_SIZE                 32768
LEVEL1_ICACHE_ASSOC                
LEVEL1_ICACHE_LINESIZE             32
LEVEL1_DCACHE_SIZE                 32768
LEVEL1_DCACHE_ASSOC                8
LEVEL1_DCACHE_LINESIZE             64
LEVEL2_CACHE_SIZE                  262144
LEVEL2_CACHE_ASSOC                 8
LEVEL2_CACHE_LINESIZE              64
LEVEL3_CACHE_SIZE                  4194304
LEVEL3_CACHE_ASSOC                 12
LEVEL3_CACHE_LINESIZE              64
LEVEL4_CACHE_SIZE                  0
LEVEL4_CACHE_ASSOC                 
LEVEL4_CACHE_LINESIZE              


Podemos apreciar que LEVEL1_DCACHE_LINESIZE =32 bytes mientras que su tamaño es 32 KiB que serian 32*1024=32768 bytes
tambien podemos apreciar su tamaño usando lscpu desde terminal


## P1.6 (0/1.0)
-Error, el tamaño total de la caché de datos es de 64 KiB, el tamaño de bloque es de 64 bytes.


7) (2.0 punto) Conociendo los datos previos, ¿es posible saber en qué momento se
produce hit o miss? De ser posible, comentar el ejemplo, de no serlo, explicar
porqué.


No,porque nos falta saber el tipo de mapeo que utiliza pero de brindarnos esta informacion, esto es posible ya que sabemos el tamaño de la cache que son de 32KiB y tambien sabemos su linesize que es de 32 bytes


## P1.7 (2/2.0)


8) (4.0 puntos) Basándose en todas las respuestas de los incisos anteriores, ¿qué
implementación genera mejores tiempos de ejecución? Justificar su respuesta.


Por el tiempo promedio calculado parece generar mejores tiempos el Programa1_2 , es decir la segunda alternativa. Esta diferencia es practicamente insignificante y esto se debe practicamente que son muy similares la unica diferencia que tienen es que en el programa numero 1 se utiliza un arreglo unidimencional antes. Hacer esto es una muy buena practica ya que ayuda a tener mejores tiempos aprovechando la localidad espacial pero lastimosamente estas mejoras significativas de tiempo son mucho mas notorias cuando el arreglo es de mas dimenciones es decir bidimencional o se emplearian mas funciones. Al no ser este el caso en un tiempo de prueba de solo 15 repeticiones podemos ver que las diferencias no son tan marcadas.

Incluso puede llegar a ser mas lento el primero ya que en las instrucciones está buscando el arreglo[i] y luego a su valor sumarlo con suma yt actualizar esta variable, mientras que en el segundo directamente al valor de suma le suma el valor de i, es decir tiene que acceder a menos posiciones de memoria.

Para programas mas complejos deberia tener mejor rendimiento el primero.


## P1.8 (0/4.0)
-Está suponiendo que el orden del arreglo es en grupos, lo cual no necesariamente es correcto. El acceso a memoria del arreglo es lo hace demorar su programa.


9) (3.0 puntos) ¿Cómo influye el tipo de datos del arreglo en este ejercicio? Esperaría
resultados similares para un tipo char, short, long? Comentar acerca de su
respuesta.

Influye bastante ya que para un arreglo o variables seguidas se tiene una posicion de memoria alejada segun el tamaño de su variable. Es decir que en la memoria cache alcanzarian mas bloques de variables si es que estas tuvieran un tipo mas pequeño. 
En el ejemplo se utiliza int para enteros que tiene un tamaño de 32 bits es decir que cada variable ocupa una posicion de memoria de 32 bits.

Al tener una memoria cache de 32 bytes line size con 32KiB si utilizar alguno de mayor tamaño ejemplo double de 64 bits = 8 bytes ocuparian mas espacio y para tipos mas grandes todabia mas, reduciria la cantidad de variables que pueden estar por bloques en la memoria cache y aumentaria los tiempos de lectura y escritura, mientras por otra parte al ser mas pequeños deberia ayudar a reducir estos tiempos ya que pueden entrar mas valores



## P1.9 (3/3.0)